### ISSUE TITLE
Add progress_bar in ComfyUI and minimum VRAM requirement.

### ISSUE THREAD

**Issue by zmwv823** [GiusTex/ComfyUI-DiffusersImageOutpaint#3]
# Add progress_bar in ComfyUI:
## pipeline_fill_sd_xl.py
```
from comfy.utils import ProgressBar
......
Add line-474: 
ComfyUI_ProgressBar = ProgressBar(int(num_inference_steps))
......
Add line-550: 
progress_bar.update()
ComfyUI_ProgressBar.update(1) # line-550
yield latents_to_rgb(latents)
......
```
# Minimum VRAM:
- With 'cpu_offload': 5.6GB.
![image](https://github.com/user-attachments/assets/301c49ad-7eb5-465f-a820-2249eb5e55d1)
![image](https://github.com/user-attachments/assets/9b74b3d9-3b3f-4062-96c1-0909dd601db1)

**Comment by GiusTex** [comment]
Sorry, what file are you referring to? `Nodes.py` ends around line 300, while you start much more later (line-474); the progress line should also be in this file, right?

**Comment by zmwv823** [comment]
> Sorry, what file are you referring to? `Nodes.py` ends around line 300, while you start much more later (line-474); the progress line should also be in this file, right? You can also open a pull request, so you can directly edit the right files

lol, file name is above the code block.
# pipeline_fill_sd_xl.py

**Comment by GiusTex** [comment]
🙈
You are completely right, omg

**Comment by GiusTex** [comment]
Done, thanks! I still thinks it would be cool if you opened a pull request, this way you would show up as a contributor

**Comment by zmwv823** [comment]
> Done, thanks! I still thinks it would be cool if you opened a pull request, this way you would show up as a contributor

LOL, not needed, it's just a slightly change.

**Comment by zmwv823** [comment]
> 🙈 You are completely right, omg

By the way diffusers folder in comfyui is `ComfyUI_windows_portable\ComfyUI\models\diffusers`.
Unet is part of diffusers model.
I also misunderstood this before.
![image](https://github.com/user-attachments/assets/c3a831f9-7612-49b4-9c59-bf9cab51f779)

**Comment by GiusTex** [comment]
And vram usage added too. I got 8,4, I don't know how you had it so low, but I reported both values

**Comment by GiusTex** [comment]
> By the way diffusers folder in comfyui is ComfyUI_windows_portable\ComfyUI\models\diffusers.

I saw that comfyui consider diffusers and unet folders the same

**Comment by GiusTex** [comment]
![image](https://github.com/user-attachments/assets/0bbc5d94-3da2-4477-8c46-54c125f9b51a)

Uhm maybe I already tried, I don't remember, it looks like both unet and diffusers folder/nodes load files, not folders

**Comment by GiusTex** [comment]
Now the choice would be whether to gather all the models (main model, vae, controlnet) in the diffusers folder, or leave them where they are, but in the end there would be no advantage in moving the files, it would be a conceptual thing

**Comment by zmwv823** [comment]
> And vram usage added too. I got 8,4, I don't know how you had it so low, but I reported both values

# Like comfyui official procedure:

## Load one model to vram to process special task, after done move this model to cpu (offload).
The order should be clip (text_encoder)--->vae (image process)--->controlnet + unet (this two model will loop use for processing in generation in Diffusers-Image-Outpaint project)--->vae (decode latents).

## It's a standard lowvram mode, but comfyui official nodes is more powerful, it can even load part of one model (such as unet) to decrease vram.

**Comment by GiusTex** [comment]
I tried loading single unet files into the pipeline

<details>
<summary>The Pipeline</summary>

```
pipe = StableDiffusionXLFillPipeline.from_pretrained(
            f"{model_path}",
            torch_dtype=torch.float16,
            vae=vae,
            controlnet=controlnet_model,
            variant="fp16"
        ).to("cuda")
```

</details>

but they didn't work, as I said comfyui seems to load only files not folders, unless I missed the right `load diffusion` etc nodes, so:
- I have to load altogheter `model`, `controlnet` and `vae`; I tried removing the vae to load it later and decode separately but it didn't work (the pipeline seems to need the vae), and I think that if I don't load it into vram I'll get a "models on different platforms" error). I also tried adding a generator in the generator to see if I can output latents, but it didn't work;
- now I removed `state_dict`, `model_file` after defining the controlnet model, instead of waiting after the generation, but for some reason it takes more vram now, I don't understand if I made something wrong somewhere but I think I reverted most of the tests;

This is the generator:

<details>
<summary>The Generator</summary>

```
(prompt_embeds,
negative_prompt_embeds,
pooled_prompt_embeds,
negative_pooled_prompt_embeds,
) = pipe.encode_prompt(final_prompt, "cuda", True)
        
generated_images = list(pipe(
prompt_embeds=prompt_embeds,
negative_prompt_embeds=negative_prompt_embeds,
pooled_prompt_embeds=pooled_prompt_embeds,
negative_pooled_prompt_embeds=negative_pooled_prompt_embeds,
image=cnet_image,
num_inference_steps=steps
))
```

</details>

and after that, I remove the model files:
```
del pipe, vae, model, controlnet_model, prompt_embeds, negative_prompt_embeds, pooled_prompt_embeds, negative_pooled_prompt_embeds 
gc.collect()
torch.cuda.empty_cache()
torch.cuda.ipc_collect()
```
but as I said for some reson now the memory doesn't get cleaned

**So...** I don't know how to get 5,6 gb of vram like you

**Comment by zmwv823** [comment]
> I tried loading single unet files into the pipeline
> 
> The Pipeline
> but they didn't work, as I said comfyui seems to load only files not folders, unless I missed the right `load diffusion` etc nodes, so:
> 
> * I have to load altogheter `model`, `controlnet` and `vae`; I tried removing the vae to load it later and decode separately but it didn't work (the pipeline seems to need the vae), and I think that if I don't load it into vram I'll get a "models on different platforms" error). I also tried adding a generator in the generator to see if I can output latents, but it didn't work;
> * now I removed `state_dict`, `model_file` after defining the controlnet model, instead of waiting after the generation, but for some reason it takes more vram now, I don't understand if I made something wrong somewhere but I think I reverted most of the tests;
> 
> This is the generator:
> 
> The Generator
> ```
> (prompt_embeds,
> negative_prompt_embeds,
> pooled_prompt_embeds,
> negative_pooled_prompt_embeds,
> ) = pipe.encode_prompt(final_prompt, "cuda", True)
>         
> generated_images = list(pipe(
> prompt_embeds=prompt_embeds,
> negative_prompt_embeds=negative_prompt_embeds,
> pooled_prompt_embeds=pooled_prompt_embeds,
> negative_pooled_prompt_embeds=negative_pooled_prompt_embeds,
> image=cnet_image,
> num_inference_steps=steps
> ))
> ```
> 
> and after that, I remove the model files:
> 
> ```
> del pipe, vae, model, controlnet_model, prompt_embeds, negative_prompt_embeds, pooled_prompt_embeds, negative_pooled_prompt_embeds 
> gc.collect()
> torch.cuda.empty_cache()
> torch.cuda.ipc_collect()
> ```
> 
> but as I said for some reson now the memory doesn't get cleaned
> 
> **So...** I don't know how to get 5,6 gb of vram like you

# Here is an article about how to optimize sdxl pipelie:
- https://www.felixsanz.dev/articles/ultimate-guide-to-optimizing-stable-diffusion-xl#model-cpu-offload

## But i don't think it's a good idea to spend too much time on it. For those who familar with comfy code (unfortunately i'm not), it can be easily intergrated into comfyui with official nodes (clip、ksampler、etc).
## For more control, the `pipeline_fill_sd_xl` need modified.
### Two input entrance: model_init-model_loader, generated_images-generation. All required inputs and process order are here, they can be modified.
- model init:
```
    def __init__( #line-80
        self,
        vae: AutoencoderKL,
        text_encoder: CLIPTextModel,
        text_encoder_2: CLIPTextModelWithProjection,
        tokenizer: CLIPTokenizer,
        tokenizer_2: CLIPTokenizer,
        unet: UNet2DConditionModel,
        controlnet: ControlNetModel_Union,
        scheduler: KarrasDiffusionSchedulers,
        force_zeros_for_empty_prompt: bool = True,
    ):
```
- generation:
```
    def __call__( #line-383
        self,
        prompt_embeds: torch.Tensor,
        negative_prompt_embeds: torch.Tensor,
        pooled_prompt_embeds: torch.Tensor,
        negative_pooled_prompt_embeds: torch.Tensor,
        image: PipelineImageInput = None,
        num_inference_steps: int = 8,
        guidance_scale: float = 1.5,
        controlnet_conditioning_scale: Union[float, List[float]] = 1.0,
        keep_model_loaded: bool = True,
    ):
```

**Comment by GiusTex** [comment]
Mine is not an sdxl pipeline, but a fill one

**Comment by GiusTex** [comment]
I already searched the past days on huggingface the sdxl pipelines and details on the fill one there aren't (or I couldn't find them), and what works for sdxl didn't work for me on sdxlfill

**Comment by GiusTex** [comment]
You can always do a pull request since I don't understand how to implement, if it works, what you say

**Comment by GiusTex** [comment]
I'll look into `model_init-model_loader`, `generated_images-generation` later, thanks for pointing them out

**Comment by zmwv823** [comment]
> You can always do a pull request since I don't understand how to implement, if it works, what you say

I'll try when it works fine and stable. And maybe the original outpiant node works for it, easy to control.
![image](https://github.com/user-attachments/assets/dc9bdebd-28a9-4ff8-9721-1356260765c1)

**Comment by GiusTex** [comment]
You can still pass image and mask to other nodes, it's just that you don't have the direction inputs, like in the outpaint space by fffiloni, or you wanted smth else?

**Comment by GiusTex** [comment]
Maybe I could add an option to choose between alignment and directions inputs

**Comment by zmwv823** [comment]
> You can still pass image and mask to other nodes, it's just that you don't have the direction inputs, like in the outpaint space by fffiloni, or you wanted smth else?

lol, not that way. 
The image padding node is easier and enough for use.

**Comment by GiusTex** [comment]
> The image padding node is easier and enough for use.

You lose the cool alignment and auto-resize options though, and the image needs to be passed to other nodes before being fed to the diffuser-outpaint node, so I don't think it's easily skippable (unless you're ok with having more nodes instead of just one)

**Comment by GiusTex** [comment]
Inside `pipeline_fill_sd_xl`,  `model_init-model_loader`/`model_init`/`model_loader` and `generated_images-generation`/`generated_images`/`generation` don't exist. As I sad, probably this pipeline is different from the normal sdxl; I'll still see if I find interesting things

**Comment by GiusTex** [comment]
After:
```
(prompt_embeds,
negative_prompt_embeds,
pooled_prompt_embeds,
negative_pooled_prompt_embeds,
) = pipe.encode_prompt(final_prompt)
```
I tried deleting:
`del pipe.text_encoder, pipe.text_encoder_2, pipe.tokenizer, pipe.tokenizer_2` (no vram change)
`del pipe.unet, pipe.control_net, pipe.vae, pipe.scheduler` (got obvious error), this means `del pipe.something` works, and since `pipe.text_encoder etc` didn't change anything I think the min vram usage is 8,3 gb, with the models I used, unless there are other ways. I tried deleting `text_encoder/s` and `tokenizer/s` even in `__init__` and `__call__` but the prior didn't changed anything and the latter gave error

**Comment by zmwv823** [comment]
# Finally with process in order, maximum vram is 5.3gb, works like comfyui lowvram mode except load whole unet into vram.
## Process order: 
- clip (text_encoders, L and G)--->vae_config for latent processing (vae model not needed, just need a int number as scale_factor (length of a tuple: block_out_channels in vae_config file (config.json---line6)))--->controlnet + unet loop--->vae decode.

## With unload all loaded model will be deleted.
## The code `device = device or self._execution_device` from gradio app isn't work in comfyui, it brings unstable bug.
## Place vae_config file and controlnet_config file into custom-node folder, in this way only single file such as `sdxl_vae_fp16_fix.safetensors  controlnet-union-sdxl-1.0-promax.safetensors` needed. 

```
from diffusers import AutoencoderKL
vae_config = os.path.join(vae_model_folder with config.json in it)
vae = AutoencoderKL.from_single_file(vae_singlefile_path, config=vae_config).to(device, dtype)
```

![image](https://github.com/user-attachments/assets/2eb9ff0d-9aea-437d-8bd2-1d717095bf37)
